### PR TITLE
#187300243 Reset user password via email

### PR DESCRIPTION
**Pull request by kanu-cast**
### What does this PR do?
This feature allows users who forgot their password to reset their passwords after confirming email

### Task description
finding user account using user email
assigning token with user id and expiry date
sending email with reset password link to user's email
verifying token validity
updating user password, hashing and saving to db

### How should this be manually tested?
Should be tested on swagger

What are the relevant pivotal tracker/Trello stories?
#187300243

#### Password reset request

Screenshot 2024-04-14 at 16 14 20

#### Password resetting

Screenshot 2024-04-14 at 16 15 09

### CODE REVIEW

**Review by ErnestTchami — APPROVED** [review]

**Review by sabatohura — APPROVED** [review]

**Review by kayigmb — APPROVED** [review]

File: src/controllers/user.controllers.ts [hunk 1]
```diff
@@ -120,6 +124,84 @@ export const userLogin = async (req: Request, res: Response) => {
...
+    const foundUser = await User.findOne({
+      where: {
+        email: req.body.email
+      }
+    });
+    if (!foundUser) return res.status(404).json({ msg: "User not found" });
+    const { firstName, id, email } = foundUser.dataValues;
+
+    const token = jwt.sign(
+      {
```
Comment (Eli250) [inline src/controllers/user.controllers.ts]: We have a function that does this operation, use it.

File: src/controllers/user.controllers.ts [hunk 2]
```diff
@@ -120,6 +124,84 @@ export const userLogin = async (req: Request, res: Response) => {
...
 };
 
+export const sendResetInstructions = async (
+  req: Request,
+  res: Response,
+  next: NextFunction
+) => {
+  try {
+    const foundUser = await User.findOne({
+      where: {
```
Comment (Eli250) [inline src/controllers/user.controllers.ts]: Use a middlelware instead of this

**Review by Eli250 — CHANGES_REQUESTED** [review]

File: src/controllers/user.controllers.ts (hunk 1 above)
Comment (kanu-cast) [reply src/controllers/user.controllers.ts]: Fixed it

File: src/controllers/user.controllers.ts (hunk 2 above)
Comment (kanu-cast) [reply src/controllers/user.controllers.ts]: fixed it

**Review by ErnestTchami — APPROVED** [review]

File: src/controllers/user.controllers.ts [hunk 3]
```diff
@@ -1,18 +1,24 @@
-import { Request, Response } from "express";
+import { NextFunction, Request, Response } from "express";
 import bcrypt from "bcrypt";
-import * as jwt from "jsonwebtoken";
+import jwt, { Jwt, JwtPayload } from "jsonwebtoken";
```
Comment (ErnestTchami) [inline src/controllers/user.controllers.ts]: Why we import jwt 2 times and also Jwtpayload ??

**Review by ErnestTchami — CHANGES_REQUESTED** [review]

File: src/controllers/user.controllers.ts (hunk 3 above)
Comment (kanu-cast) [reply src/controllers/user.controllers.ts]: fixed it

**Review by ErnestTchami — APPROVED** [review]

**Review by istafrica-dennw — APPROVED** [review]

**Pull request merged**